src/app.test.ts

```diff
@@ -1,51 +1,267 @@
 
 tmp.setGracefulCleanup()
 
-beforeEach((done, t) => {
+type TestContext = {
+  accessDb: (cb: (db: DatabaseInstance) => void) => void
+  server: FastifyInstance
+  sampleTileJSON: TileJSON
+}
+
+function assertSampleTileJSONIsValid(data: unknown): asserts data is TileJSON {
+  if (!validateTileJSON(data)) {
+    const message = `Sample input does not conform to TileJSON schema spec: ${JSON.stringify(
+      validateTileJSON.errors,
+      null,
+      2
+    )}`
+
+    throw new Error(message)
+  }
+}
+
+before(() => {
+  // Check if prisma/migrations directory exists in project
+  if (!fs.existsSync(path.resolve(__dirname, '../prisma/migrations'))) {
+    throw new Error(
+      'Could not find prisma migrations directory. Make sure you run `npm run prisma:migrate-dev -- --name MIGRATION_NAME_HERE` first!'
+    )
+  }
+
+  assertSampleTileJSONIsValid(mapboxRasterTilejson)
+
+  mockTileServer.listen()
+})
+
+beforeEach((t) => {
   const { name: dataDir } = tmp.dirSync({ unsafeCleanup: true })
-  t.context.app = app({ logger: false }, { dataDir })
-  done()
+
+  const dbPath = path.resolve(dataDir, 'test.db')
+
+  t.context = {
+    accessDb: (cb: (db: DatabaseInstance) => void) => {
+      const db = new DB(dbPath, { readonly: true })
+      cb(db)
+      db.close()
+    },
+    server: app({ logger: false }, { dbPath }),
+    sampleTileJSON: mapboxRasterTilejson,
+  }
+})
+
+afterEach((t) => {
+  t.context.server.close()
+  mockTileServer.resetHandlers()
 })
 
-afterEach((done, t) => {
-  t.context.app.close().then(done)
+teardown(() => {
+  mockTileServer.close()
 })
 
 test('GET /tilesets (empty)', async (t) => {
-  const { app } = t.context
+  const { server } = t.context as TestContext
+
+  const response = await server.inject({ method: 'GET', url: '/tilesets' })
+
+  t.equal(response.statusCode, 200, 'returns a status code of 200')
 
-  const response = await app.inject({ method: 'GET', url: '/tilesets' })
-  t.strictEqual(response.statusCode, 200, 'returns a status code of 200')
-  t.strictEqual(
+  t.equal(
     response.headers['content-type'],
     'application/json; charset=utf-8',
     'returns correct content-type header'
   )
-  t.deepEqual(response.json(), [], 'returns empty array')
+
+  t.same(response.json(), [], 'returns empty array')
+
+  t.end()
+})
+
+test('GET /tilesets (not empty)', async (t) => {
+  const { sampleTileJSON, server } = t.context as TestContext
+
+  await server.inject({
+    method: 'POST',
+    url: '/tilesets',
+    payload: sampleTileJSON,
+  })
+
+  const expectedId = getTilesetId(sampleTileJSON)
+  const expectedTileUrl = `http://localhost:80/tilesets/${expectedId}/{z}/{x}/{y}`
+  const expectedResponse = [
+    {
+      ...sampleTileJSON,
+      id: expectedId,
+      tiles: [expectedTileUrl],
+    },
+  ]
+
+  const response = await server.inject({ method: 'GET', url: '/tilesets' })
+
+  t.same(response.json(), expectedResponse)
+
+  t.end()
 })
 
 test('POST /tilesets', async (t) => {
```

Comment:
I think this test is testing the same thing as the previous test, but it's also relying on internal implementation details (accessdb). I think we should just be testing that posting a tile set makes it available to GET

Reply:
is the `accessDb` functionality not necessary in general? was debating if that would be useful for verifying data but can see how it's a little overreaching

Reply:
I would drop it. It's better to verify everything using only the public API if possible, avoiding reaching into the internals. What we actually want to test is "if I put something in here, can I get it out again", not "if I put something in here, is it somewhere in the internals (but not necessarily accessible)"

Reply:
makes sense. will get rid of it

Reply:
Similarly I would drop the `getTilesetId` function from the tests. Hardcode the values, and treat it essentially as a manual snapshot test for this value (e.g. you copy the value and manually put it in the test, at the same time manually verifying it looks right (e.g. like a hash).